### ISSUE TITLE
move `cancel` to job

### ISSUE THREAD

**Issue by andre-merzky** [ExaWorks/job-api-spec#74]
I suggest to move `cancel` to the Job  class.  At the moment, `wait` is on the Job, and `cancel` is on the Executor - this is inconsistent, and I frequently forget which is where.

**Comment by hategan** [comment]
That's all great, but it means the job must keep a reference to the executor, which I'd like to avoid.

In principle:
- executor: does things (submit, cancel, etc.) that are specific to implementation
- job: has things (e.g., spec, states)

waitFor() is in Job because it's not really as much a "do thing" as it is a fancy form of "state == something" and is not specific to implementations (e.g. PBS, SLURM, etc.)

**Comment by andre-merzky** [comment]
> That's all great, but it means the job must keep a reference to the executor, which I'd like to avoid.

How is that different from wait?

**Comment by andre-merzky** [comment]
That's all great, but as said, I think it makes the API inconsistent.  The goal was to have a nice API, not a simple implementation (as we clearly state in the document).

**Comment by hategan** [comment]
> I think it makes the API inconsistent.

Ok, but can we try to ground the discussion in something else than personal views?

**Comment by andre-merzky** [comment]
> > I think it makes the API inconsistent.
> 
> Ok, but can we try to ground the discussion in something else than personal views?

It is my personal view I guess, but then again it is also not really subjective.  Can we try try to ground the discussion in something else than implementation issues? :-)  Sorry for the cheap shot, it was just offering itself... (**_\*blush\*_**)

**Comment by hategan** [comment]
> It is my personal view I guess, but then again it is also not really subjective.

I don't doubt that. But let's try to articulate the "not really" part of "subjective" (aka. the objective). I could try to guess, but I'm fairly sure the result would not be reliable.

**Comment by andre-merzky** [comment]
> But let's try to articulate the "not really" part of "subjective" (aka. the objective).

I tried to do that, let me try to phrase it clearer / more explicit.  

We have an Executor class to represent an LRM (or other executors) which manage jobs, and a Job class to  represent those managed jobs as stateful entities.  `executor.submit` is used to relate Jobs to Executors, and from that point on we expect the job state to be managed by the executor.  After that submit, there are two more methods defined which relate to job state (apart from inspection of said state which is always possible): `wait` and `cancel`.  It is inconsistent that one of those methods is defined on the Executor, and the other on the Job.  I suggest to define both on the Job.  

Defining both on the Executor would also be consistent, but I _subjectively_ like the other option more.

**Comment by hategan** [comment]
I see.

`wait()` and `cancel()` are not on the same footing though. Again, `submit()` and `cancel()` do things to the job. In other words, they can change the state of a job. `wait()` does not do things to the job (i.e., it cannot change its state).

I guess we could add `JobExecutor.wait(job, ...)` given that we also have multi-job callbacks on the executor. But I'd suggest keeping `Job.wait()` since, even conceptually, waiting on a state is independent of how the jobs are handled concretely (i.e. LRM).

**Comment by SteVwonder** [comment]
>  it means the job must keep a reference to the executor, which I'd like to avoid.

That's a valid point that each job would need to add a reference to the executor that submitted it, but that wouldn't be hard to achieve.  Using python for a straw man example, it could be implemented with just a few lines:

```python
class Executor:
    def submit(self, job):
        ...
        job._executor = self
        ...
class Job:
    def cancel(self):
        self._executor.cancel(self)
``` 

Could you expand on why you want to avoid the job keeping a reference to its executor?  Is it a memory overhead concern?

**Comment by hategan** [comment]
> Could you expand on why you want to avoid the job keeping a reference to its executor? Is it a memory overhead concern?

Because it's a structural change done with the sole purpose of implementing this one method. You may also note that your example does not alleviate the need for `cancel()` to be in the executor.

**Comment by hategan** [comment]
> You may also note that your example does not alleviate the need for `cancel()` to be in the executor.

Although that is a consequence of me declaring that you need to hold a reference to the executor, which assumes that all implementation (i.e. PBS/SLURM) specific details are held by the executor.

**Comment by andre-merzky** [comment]
The issue was about an API consistency, the discussion is about possibly keeping a reference in an object.  I'd like to suggest to refocus on the API consistency part.

**Comment by SteVwonder** [comment]
> I guess we could add JobExecutor.wait(job, ...) given that we also have multi-job callbacks on the executor.
> You may also note that your example does not alleviate the need for cancel() to be in the executor.

Can they be in both places as a convenience to users?  I get what you are saying about the executor being the "doer" of actions on jobs and the job having "things" that can be inspected or waited on.  That distinction makes sense, but if a user has a reference to a `Job` and they want to cancel it, it seems natural to just want to do `job.cancel()`.  That form also alleviates the burden from the user of having to keep track of which jobs were submitted by which executors in order to cancel them.  If it was a huge ordeal to implement within J/PSI, I'd be with you, but this doesn't seem like that much complexity or effort on our side for a nice ergonomic gain on the user side.

**Comment by andre-merzky** [comment]
This is resolved via #100
